src/components/utils/caret.ts

```diff
@@ -0,0 +1,188 @@
+   *
+   * <div>|hello</div> <---- Selection references to <div> instead of text node
+   *
+   *
+   */
+  if (focusNode.nodeType !== Node.TEXT_NODE && focusNode.childNodes.length > 0) {
+    /**
+     * In normal cases, focusOffset is a child index.
+     */
+    if (focusNode.childNodes[focusOffset]) {
+      focusNode = focusNode.childNodes[focusOffset];
+      focusOffset = 0;
+    /**
+     * But in Firefox, focusOffset can be 1 with the single child.
+     */
+    } else {
+      focusNode = focusNode.childNodes[focusOffset - 1];
+      focusOffset = focusNode.textContent.length;
+    }
+  }
+
+  return [focusNode, focusOffset];
+}
+
+/**
+ * Checks content at left or right of the passed node for emptiness.
+ *
+ * @param contenteditable - The contenteditable element containing the nodes.
+ * @param fromNode - The starting node to check from.
+ * @param offsetInsideNode - The offset inside the starting node.
+ * @param direction - The direction to check ('left' or 'right').
+ * @returns true if adjacent content is empty, false otherwise.
+ */
+export function checkContenteditableSliceForEmptiness(contenteditable: HTMLElement, fromNode: Node, offsetInsideNode: number, direction: 'left' | 'right'): boolean {
+  const range = document.createRange();
+
+  /**
+   * In case of "left":
+   * Set range from the start of the contenteditable to the passed offset
+   */
+  if (direction === 'left') {
+    range.setStart(contenteditable, 0);
+    range.setEnd(fromNode, offsetInsideNode);
+
+  /**
+   * In case of "right":
+   * Set range from the passed offset to the end of the contenteditable
+   */
+  } else {
+    range.setStart(fromNode, offsetInsideNode);
+    range.setEnd(contenteditable, contenteditable.childNodes.length);
+  }
+
+  /**
+   * Clone the range's content and check its text content
+   */
+  const clonedContent = range.cloneContents();
+  const tempDiv = document.createElement('div');
+
+  tempDiv.appendChild(clonedContent);
+
+  const textContent = tempDiv.textContent || '';
+
+  /**
+   * In HTML there are two types of whitespaces:
+   * - visible (&nbsp;)
+   * - invisible (trailing spaces, tabs, etc.)
+   *
+   * If text contains only invisible whitespaces, it is considered to be empty
+   */
+  return isCollapsedWhitespaces(textContent);
+}
+
+/**
+ * Checks if caret is at the start of the passed input
+ *
+ * Cases:
+ *  Native input:
+ *   - if offset is 0, caret is at the start
+ *  Contenteditable:
+ *   - caret at the first text node and offset is 0 — caret is at the start
+ *   - caret not at the first text node — we need to check left siblings for emptiness
+ *   - caret offset > 0, but all left part is visible (nbsp) — caret is not at the start
+ *   - caret offset > 0, but all left part is invisible (whitespaces) — caret is at the start
+ *
+ * @param input - input where caret should be checked
+ */
+export function isAtStartOfInput(input: HTMLElement): boolean {
```

Comment:
This function name is a bit confusing, hard to understand what exactly is 'at start of input'. 
I suggest calling it 'isCaretAtStartOfInput'